internal/graph/singleflight_resolver_test.go

```diff
@@ -0,0 +1,174 @@
+
+	checker := NewLocalChecker(
+		storagewrappers.NewCombinedTupleReader(ds, []*openfgav1.TupleKey{}),
+	)
+	defer checker.Close()
+
+	resp, err := checker.ResolveCheck(ctx, &ResolveCheckRequest{
+		StoreID:            storeID,
+		TupleKey:           tuple.NewTupleKey("folder:3", "viewer", "user:jon"),
+		ContextualTuples:   nil,
+		ResolutionMetadata: &ResolutionMetadata{Depth: defaultResolveNodeLimit},
+	})
+
+	require.NoError(t, err)
+	require.True(t, resp.GetAllowed())
+	require.GreaterOrEqual(t, resp.GetResolutionMetadata().DatastoreQueryCount, uint32(4))
+	require.LessOrEqual(t, resp.GetResolutionMetadata().DatastoreQueryCount, uint32(7))
+}
+func TestSingleflightResolverVersusWithout(t *testing.T) {
+	ds := memory.New()
+	defer ds.Close()
+
+	storeID := ulid.Make().String()
+
+	var tuples = []*openfgav1.TupleKey{
+		tuple.NewTupleKey("folder:1", "viewer", "user:jon"),
+	}
+	for i := 1; i <= 15; i++ {
+		tuples = append(tuples, tuple.NewTupleKey(fmt.Sprintf("folder:%d", i+1), "parent1", fmt.Sprintf("folder:%d", i)))
+		tuples = append(tuples, tuple.NewTupleKey(fmt.Sprintf("folder:%d", i+1), "parent2", fmt.Sprintf("folder:%d", i)))
+		tuples = append(tuples, tuple.NewTupleKey(fmt.Sprintf("folder:%d", i+1), "parent3", fmt.Sprintf("folder:%d", i)))
+	}
+
+	err := ds.Write(context.Background(), storeID, nil, tuples)
+	require.NoError(t, err)
+
+	model := testutils.MustTransformDSLToProtoWithID(`model  
+    schema 1.1  
+
+type user  
+
+type folder  
+    relations  
+    define parent1: [folder]  
+    define parent2: [folder]  
+    define parent3: [folder]  
+    define viewer: [user] or viewer from parent1 or viewer from parent2 or viewer from parent3`)
+
+	ctx := typesystem.ContextWithTypesystem(context.Background(), typesystem.New(model))
+
+	checkReq := ResolveCheckRequest{
+		StoreID:            storeID,
+		TupleKey:           tuple.NewTupleKey("folder:10", "viewer", "user:jon"),
+		ContextualTuples:   nil,
+		ResolutionMetadata: &ResolutionMetadata{Depth: defaultResolveNodeLimit},
+	}
+
+	checkerWithoutSingleflight := NewLocalChecker(
+		storagewrappers.NewCombinedTupleReader(ds, []*openfgav1.TupleKey{}),
+	)
+	defer checkerWithoutSingleflight.Close()
+
+	req := checkReq
+	respWithoutSingleflight, err := checkerWithoutSingleflight.ResolveCheck(ctx, &req)
+
+	require.NoError(t, err)
+	require.True(t, respWithoutSingleflight.GetAllowed())
+
+	checkerWithSingleflight := NewLocalChecker(
+		storagewrappers.NewCombinedTupleReader(ds, []*openfgav1.TupleKey{}),
+		WithSingleflightResolver(),
+	)
+	defer checkerWithSingleflight.Close()
+
+	req = checkReq
+	resWithSingleflight, err := checkerWithSingleflight.ResolveCheck(ctx, &req)
+
+	require.NoError(t, err)
+	require.True(t, resWithSingleflight.GetAllowed())
+	require.LessOrEqual(t, resWithSingleflight.GetResolutionMetadata().DatastoreQueryCount, respWithoutSingleflight.GetResolutionMetadata().DatastoreQueryCount)
+}
+
+func TestSingleflightResolverWithCycle(t *testing.T) {
```

Comment:
Any chance we can structure these tests in "given, when thens"? Quite hard to navigate them IMO.

Reply:
Can you elaborate what is hard to navigate? Theses test are laid out in a way you describe – given a model and tuples that produce parallel evaluations, we expect a correct resolution and fewer DB queries.

Reply:
So the tests might be laid out in that way logically, but it's hard to navigate and writing these tests doesn't seem easy and intuitive. 

What I'd suggest is something like the following:

```go
func TestSingleflightResolver(t *testing.T) {
	t.Parallel()
	
	t.Run("it can correctly resolve checks with fewer data store hits due to deduplication", func(t *testing.T) {
		model := `
model
schema 1.1

type user

type folder
relations
  define parent1: [folder]
  define parent2: [folder]
  define viewer: [user] or viewer from parent1 or viewer from parent2 or viewer
`
		
		tuples := []*openfgav1.TupleKey{
			tuple.NewTupleKey("folder:1", "viewer", "user:jon"),

			tuple.NewTupleKey("folder:2", "parent1", "folder:1"),
			tuple.NewTupleKey("folder:2", "parent2", "folder:1"),

			tuple.NewTupleKey("folder:3", "parent1", "folder:2"),
			tuple.NewTupleKey("folder:3", "parent2", "folder:2"),
		}
		
		dataStore, storeID := givenAnInMemoryDataStoreWithModelAndTuples(t, model, tuples)
		
		response, err := whenIResolveTheFollowingCheck(t, storeID, tuple)
		
		require.NoError(t, err)
		assert.True(t, response.GetAllowed())
		assert.GreaterOrEqual(t, response.GetResolutionMetadata().DatastoreQueryCount, uint32(4))
		assert.LessOrEqual(t, response.GetResolutionMetadata().DatastoreQueryCount, uint32(7))
	})
	
	t.Run("another test case", func(t *testing.T) {
		// given
		// when
		// then
	})
}
```

The above makes it so much easier to keep on writing tests with various combinations and I can read them like a book.

Reply:
https://github.com/cucumber/godog would lend itself incredibly well within this project, happy to set it up if y'all agree

Reply:
@sergiught definitely open to looking into alternative ways to make our tests more sexy, but that's well beyond the scope of this PR. As for `godog` that would be interesting, but at first glance it's confusing as hell as compared to declarative style test definitions. Maybe we can save that for a hackathon week and see how the team likes it?

As for the suggestion, we don't have any other test cases contained within the test. The test _is_ the test case. The name captures the intent of what is being tested, and so from that perspective I think it is fine as is.